File: ocaml/11-20/19.ml
(* Rotate a list N places to the left. (medium) *)

let rec take n xs = match (n, xs) with
  | (_, []) | (0, _) -> []
  | (_, y::ys) -> y::(take (n-1) ys)

let rec skip n xs = match (n, xs) with
  | (_, []) | (0, _) -> xs
  | (_, y::ys) -> skip (n-1) ys

let rec rotate xs n =
  if n < 0 then rotate xs ((List.length xs) + n)
  else (skip n xs) @ (take n xs)


let _ = assert ((rotate ["a"; "b"; "c"; "d"; "e"; "f"; "g"; "h"] 3) = ["d"; "e"; "f"; "g"; "h"; "a"; "b"; "c"])
let _ = assert ((rotate ["a"; "b"; "c"; "d"; "e"; "f"; "g"; "h"] (-2)) = ["g"; "h"; "a"; "b"; "c"; "d"; "e"; "f"])

File: prolog/11-20/19.pl
/*
Rotate a list N places to the left.
    Examples:
    ?- rotate([a,b,c,d,e,f,g,h],3,X).
    X = [d,e,f,g,h,a,b,c]

    ?- rotate([a,b,c,d,e,f,g,h],-2,X).
    X = [g,h,a,b,c,d,e,f]

    Hint: Use the predefined predicates length/2 and append/3, as well as the result of problem 1.17.
*/

:- include('17').

rotate([], _, []).
rotate(Xs, N, Zs) :- 
  N >= 0,
  take(N, Xs, [], NewEnd),
  drop(N, Xs, NewBegin),
  append(NewBegin, NewEnd, Zs).
  
rotate(Xs, N, Zs) :-
  N < 0,
  length(Xs, L),
  N1 is L + N,
  rotate(Xs, N1, Zs).
